// zip-lib/lib/unzip.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Unzip = void 0;
const node_fs_1 = require("node:fs");
const fs = require("node:fs/promises");
const path = require("node:path");
const promises_1 = require("node:stream/promises");
const yauzl = require("yauzl");
const cancelable_1 = require("./cancelable");
const exfs = require("./fs");
class EntryEvent {
    /**
     *
     */
    constructor(_entryCount) {
        this._entryCount = _entryCount;
        this._isPrevented = false;
    }
    get entryName() {
        return this._entryName;
    }
    set entryName(name) {
        this._entryName = name;
    }
    get entryCount() {
        return this._entryCount;
    }
    get isPrevented() {
        return this._isPrevented;
    }
    preventDefault() {
        this._isPrevented = true;
    }
    reset() {
        this._isPrevented = false;
    }
}
class EntryContext {
    constructor(_targetFolder, _realTargetFolder, symlinkAsFileOnWindows) {
        this._targetFolder = _targetFolder;
        this._realTargetFolder = _realTargetFolder;
        this.symlinkAsFileOnWindows = symlinkAsFileOnWindows;
        this._symlinkFileNames = [];
        this._symlinkFolders = [];
        this._inspectedFolders = [];
        this._ensuredFolders = [];
    }
    get decodeEntryFileName() {
        return this._decodeEntryFileName;
    }
    set decodeEntryFileName(name) {
        this._decodeEntryFileName = name;
    }
    get targetFolder() {
        return this._targetFolder;
    }
    get realTargetFolder() {
        return this._realTargetFolder;
    }
    get symlinkFileNames() {
        return this._symlinkFileNames;
    }
    addSymlinkFolder(folder, realpath) {
        const find = this._symlinkFolders.find((item) => item.folder === folder);
        if (!find) {
            this._symlinkFolders.push({ folder, realpath });
        }
    }
    getFilePath() {
        return path.resolve(path.join(this.targetFolder, this.decodeEntryFileName));
    }
    async inspectFolder(folder) {
        if (this._inspectedFolders.includes(folder) || folder === path.dirname(folder)) {
            return;
        }
        await this.inspectFolder(path.dirname(folder));
        const folderStat = await exfs.statFolder(folder);
        if (!folderStat) {
            return;
        }
        if (folderStat.isSymbolicLink) {
            this.addSymlinkFolder(folder, folderStat.realpath);
        }
        this._inspectedFolders.push(folder);
    }
    async ensureFolder(folder) {
        if (this._ensuredFolders.includes(folder)) {
            return;
        }
        await this.inspectFolder(path.dirname(folder));
        const folderStat = await exfs.ensureFolder(folder);
        if (folderStat.isSymbolicLink) {
            this.addSymlinkFolder(folder, folderStat.realpath);
        }
        this._ensuredFolders.push(folder);
    }
    isSymlinkTargetOutsideTargetFolder(linkTarget, linkFilePath) {
        const fileDir = path.dirname(linkFilePath);
        return exfs.isOutside(this.realTargetFolder, path.resolve(fileDir, linkTarget));
    }
    async isOutsideTargetFolder(tpath) {
        if (this._symlinkFileNames.length === 0 && this._symlinkFolders.length === 0) {
            return false;
        }
        if (process.platform === "win32" && this.symlinkAsFileOnWindows) {
            return false;
        }
        for (const { folder, realpath } of this._symlinkFolders) {
            if (tpath.includes(folder)) {
                if (exfs.isOutside(this.realTargetFolder, realpath)) {
                    return true;
                }
            }
        }
        for (const fileName of this._symlinkFileNames) {
            if (tpath.includes(fileName)) {
                const realFilePath = await exfs.realpath(tpath);
                if (exfs.isOutside(this.realTargetFolder, realFilePath)) {
                    return true;
                }
            }
        }
        return false;
    }
}
/**
 * Extract the zip file.
 */
class Unzip extends cancelable_1.Cancelable {
    /**
     *
     */
    constructor(options) {
        super();
        this.options = options;
    }
    async extract(zipFileOrBuffer, targetFolder) {
        const token = new cancelable_1.CancellationToken();
        this.token = token;
        try {
            const { zfile, realTargetFolder } = await this.prepareExtraction(zipFileOrBuffer, targetFolder, token);
            this.zipFile = zfile;
            await this.processEntries(zfile, targetFolder, realTargetFolder, token);
        }
        finally {
            if (this.token === token) {
                this.token = null;
            }
        }
    }
    async prepareExtraction(zipFileOrBuffer, targetFolder, token) {
        if (this.isOverwrite()) {
            await exfs.rimraf(targetFolder);
        }
        if (token.isCancelled) {
            throw this.canceledError();
        }
        await exfs.ensureFolder(targetFolder);
        const realTargetFolder = await exfs.realpath(targetFolder);
        const zfile = await this.openZip(zipFileOrBuffer, token);
        return { zfile, realTargetFolder };
    }
    async processEntries(zfile, targetFolder, realTargetFolder, token) {
        return await new Promise((resolve, reject) => {
            let extractedEntriesCount = 0;
            let anyError = null;
            const total = zfile.entryCount;
            const entryContext = new EntryContext(targetFolder, realTargetFolder, this.symlinkToFile());
            const entryEvent = new EntryEvent(total);
            const settle = this.createPromiseSettler(resolve, reject);
            const disposeCancel = token.onCancelled(() => {
                this.closeZip(zfile);
                settle.reject(this.canceledError());
            });
            zfile.once("error", (err) => {
                disposeCancel();
                anyError = this.wrapError(err, token.isCancelled);
                this.closeZip(zfile);
                settle.reject(anyError);
            });
            zfile.once("close", () => {
                disposeCancel();
                if (this.zipFile === zfile) {
                    this.zipFile = null;
                }
                if (anyError) {
                    settle.reject(this.wrapError(anyError, token.isCancelled));
                }
                else if (token.isCancelled) {
                    settle.reject(this.canceledError());
                }
                else if (extractedEntriesCount >= total) {
                    // If the zip content is empty, it will not receive the zfile.on("entry") event.
                    settle.resolve();
                }
            });
            zfile.on("entry", async (entry) => {
                try {
                    await this.handleZipEntry(zfile, entry, entryContext, entryEvent, token);
                    extractedEntriesCount++;
                    if (extractedEntriesCount >= total) {
                        settle.resolve();
                    }
                }
                catch (error) {
                    anyError = this.wrapError(error, token.isCancelled);
                    this.closeZip(zfile);
                    settle.reject(anyError);
                }
            });
            this.readNextEntry(zfile, token);
        });
    }
    readNextEntry(zfile, token) {
        if (token.isCancelled) {
            this.closeZip(zfile);
            return;
        }
        zfile.readEntry();
    }
    createPromiseSettler(resolve, reject) {
        let settled = false;
        return {
            resolve: () => {
                if (settled) {
                    return;
                }
                settled = true;
                resolve();
            },
            reject: (error) => {
                if (settled) {
                    return;
                }
                settled = true;
                reject(error);
            },
        };
    }
    async handleZipEntry(zfile, entry, entryContext, entryEvent, token) {
        // use UTF-8 in all situations
        // see https://github.com/thejoshwolfe/yauzl/issues/84
        const rawName = entry.fileName.toString("utf8");
        // allow backslash
        const fileName = rawName.replace(/\\/g, "/");
        // Because decodeStrings is false, we need to manually verify the entry name
        // see https://github.com/thejoshwolfe/yauzl#validatefilenamefilename
        const errorMessage = yauzl.validateFileName(fileName);
        if (errorMessage != null) {
            throw new Error(errorMessage);
        }
        entryEvent.entryName = fileName;
        this.onEntryCallback(entryEvent);
        entryContext.decodeEntryFileName = fileName;
        if (entryEvent.isPrevented) {
            entryEvent.reset();
            this.readNextEntry(zfile, token);
            return;
        }
        await this.handleEntry(zfile, entry, entryContext, token);
    }
    /**
     * Cancel extraction.
     * If the cancel method is called after the extract is complete, nothing will happen.
     */
    cancel() {
        if (this.token) {
            this.token.cancel();
            this.token = null;
        }
        this.closeZip();
    }
    closeZip(zfile) {
        const target = zfile !== null && zfile !== void 0 ? zfile : this.zipFile;
        if (target) {
            target.close();
        }
        if (!zfile || this.zipFile === zfile) {
            this.zipFile = null;
        }
    }
    async openZip(zipFileOrBuffer, token) {
        const options = {
            lazyEntries: true,
            // see https://github.com/thejoshwolfe/yauzl/issues/84
            decodeStrings: false,
        };
        try {
            return await new Promise((resolve, reject) => {
                const callback = (err, zfile) => {
                    if (err) {
                        reject(this.wrapError(err, token.isCancelled));
                    }
                    else {
                        resolve(zfile);
                    }
                };
                if (typeof zipFileOrBuffer === "string") {
                    yauzl.open(zipFileOrBuffer, options, callback);
                }
                else {
                    yauzl.fromBuffer(zipFileOrBuffer, options, callback);
                }
            });
        }
        catch (error) {
            throw this.wrapError(error, token.isCancelled);
        }
    }
    async handleEntry(zfile, entry, entryContext, token) {
        if (/\/$/.test(entryContext.decodeEntryFileName)) {
            // Directory file names end with '/'.
            // Note that entries for directories themselves are optional.
            // An entry's fileName implicitly requires its parent directories to exist.
            await exfs.ensureFolder(entryContext.getFilePath());
            this.readNextEntry(zfile, token);
        }
        else {
            // file entry
            await this.extractEntry(zfile, entry, entryContext, token);
        }
    }
    openZipFileStream(zfile, entry, token) {
        return new Promise((resolve, reject) => {
            zfile.openReadStream(entry, (err, readStream) => {
                if (err) {
                    reject(this.wrapError(err, token.isCancelled));
                }
                else {
                    resolve(readStream);
                }
            });
        });
    }
    async extractEntry(zfile, entry, entryContext, token) {
        const filePath = entryContext.getFilePath();
        const fileDir = path.dirname(filePath);
        await entryContext.inspectFolder(fileDir);
        const outside = await entryContext.isOutsideTargetFolder(fileDir);
        if (outside) {
            const error = new Error(`Refuse to write file outside "${entryContext.targetFolder}", file: "${filePath}"`);
            error.name = "AFWRITE";
            return Promise.reject(error);
        }
        const readStream = await this.openZipFileStream(zfile, entry, token);
        await this.writeEntryToFile(readStream, entry, entryContext, token);
        this.readNextEntry(zfile, token);
    }
    async writeEntryToFile(readStream, entry, entryContext, token) {
        var _a;
        try {
            const filePath = entryContext.getFilePath();
            const mode = this.modeFromEntry(entry);
            // see https://unix.stackexchange.com/questions/193465/what-file-mode-is-a-symlink
            const isSymlink = (mode & 0o170000) === 0o120000;
            if (isSymlink) {
                entryContext.symlinkFileNames.push(path.resolve(path.join(entryContext.targetFolder, entryContext.decodeEntryFileName)));
            }
            if (isSymlink && !this.symlinkToFile()) {
                const linkContent = await this.readStreamContent(readStream, token);
                if (((_a = this.options) === null || _a === void 0 ? void 0 : _a.safeSymlinksOnly) &&
                    entryContext.isSymlinkTargetOutsideTargetFolder(linkContent, filePath)) {
                    const error = new Error(`Dangerous link path was refused : "${entryContext.targetFolder}", file: "${filePath}", target: "${linkContent}". Set safeSymlinksOnly to false to allow writing through this symlink.`);
                    error.name = "AF_ILLEGAL_TARGET";
                    throw error;
                }
                await entryContext.ensureFolder(path.dirname(filePath));
                await this.createSymlink(linkContent, filePath);
            }
            else {
                await entryContext.ensureFolder(path.dirname(filePath));
                const fileStream = (0, node_fs_1.createWriteStream)(filePath, { mode });
                const pipelinePromise = (0, promises_1.pipeline)(readStream, fileStream);
                const disposeCancel = token.onCancelled(() => {
                    fileStream.destroy(this.canceledError());
                });
                try {
                    await pipelinePromise;
                }
                finally {
                    disposeCancel();
                }
            }
        }
        catch (error) {
            throw this.wrapError(error, token.isCancelled);
        }
    }
    async readStreamContent(readStream, token) {
        let linkContent = "";
        const readPromise = new Promise((resolve, reject) => {
            readStream.on("data", (chunk) => {
                linkContent += typeof chunk === "string" ? chunk : chunk.toString();
            });
            readStream.once("end", resolve);
            readStream.once("error", (err) => {
                reject(this.wrapError(err, token.isCancelled));
            });
        });
        const disposeCancel = token.onCancelled(() => {
            readStream.destroy(this.canceledError());
        });
        try {
            await readPromise;
            return linkContent;
        }
        finally {
            disposeCancel();
        }
    }
    modeFromEntry(entry) {
        const attr = entry.externalFileAttributes >> 16 || 33188;
        return [448 /* S_IRWXU */, 56 /* S_IRWXG */, 7 /* S_IRWXO */]
            .map((mask) => attr & mask)
            .reduce((a, b) => a + b, attr & 61440 /* S_IFMT */);
    }
    async createSymlink(linkContent, filePath) {
        let linkType = "file";
        if (process.platform === "win32") {
            if (/\/$/.test(linkContent)) {
                linkType = "dir";
            }
            else {
                let targetPath = linkContent;
                if (!path.isAbsolute(linkContent)) {
                    targetPath = path.join(path.dirname(filePath), linkContent);
                }
                try {
                    const stat = await fs.stat(targetPath);
                    if (stat.isDirectory()) {
                        linkType = "dir";
                    }
                }
                catch (_error) {
                    // ignore
                }
            }
        }
        await fs.symlink(linkContent, filePath, linkType);
    }
    isOverwrite() {
        var _a;
        if ((_a = this.options) === null || _a === void 0 ? void 0 : _a.overwrite) {
            return true;
        }
        return false;
    }
    onEntryCallback(event) {
        var _a;
        if ((_a = this.options) === null || _a === void 0 ? void 0 : _a.onEntry) {
            this.options.onEntry(event);
        }
    }
    symlinkToFile() {
        let symlinkToFile = false;
        if (process.platform === "win32") {
            if (this.options && this.options.symlinkAsFileOnWindows === false) {
                symlinkToFile = false;
            }
            else {
                symlinkToFile = true;
            }
        }
        return symlinkToFile;
    }
}
exports.Unzip = Unzip;
